# Prediction of CO2 emissions from country-specific data
# A Machine Learning project 
### by Vladislav Todorov

***

# Stage 1: Data cleaning and preparation

***

### Notebook Contents:
0. Introduction - project and notebook summary, notes on the data source
1. Notebook setup - libraries and data import
2. Global data overview
3. Definition of the initial project goals
3. Data cleaning
    - dealing with missing values
    - transformation of the columns into a numerical data type
    - renaming of features
    - removing empty columns and rows
4. Data frame transformation
    - melting of the data for each variable
    - integration of the data into a suitable data frame format
5. Removal of missing values
    - detection of missing values
    - removal of missing values by filtering the columns and rows, so that minimal amount of features and rows are lost
5. Export the clean data frame to a file

***

## 0. Introduction

### Project summary
**Aim of the project**:
Analysis of country-specific data and development of machine learning models in order to predict CO2 emissions from country parameters. The project uses the publicly available dataset Climate Change Data from the World Bank Group, which provides data on the vast majority of countries over a range of years for parameters such as:

* country: the vast majority of countries worldwide
* year: ranging from 1990 to 2011
* various emissions of greenhouse gases such as CO2, CH4, N2O, others
* population-specific parameters: population count, urban population, population growth, etc.
* country economic indicators: GDP, GNI, Foreign Direct Investment, etc.
* land-related parameters: cereal yield, agricultural land, Nationally terrestrial protected areas, etc.
* climate data: precipitations, national disasters, etc.
* energy use
* counts of certain types of medical personnel
* etc.

The project is divided into three stages:

1. Data cleaning and preparation
2. Data exploration and visualization
3. Predictive analysis with the Random Forest machine learning algorithm

Each of the stages is described in a separate Jupyter Notebook(.ipynp file) and a derived pdf file.

***

### Notebook summary - Stage 1: Data cleaning and preparation
**Aim of this notebook**: The subject of this particular notebook is to explain the first stage of the project - the cleaning and transformation of the available data in order to prepare it for the visualization and analysis (described in further notebooks).

**Input**: excel data file from the original online source

**Output**: comma separated values (CSV) file containing the cleaned data ready for visualization and analysis

**Programming language**: Python 3.7

**Libraries used in this notebook**: pandas, numpy

***

### Data source

The used data comes from the Climate Change Data of the World Bank Group, which provides country-specific data on parameters such as CO2 emissions, energy use, population count, urban population, cereal yield, nationally terrestrial protected areas, GDP, GNI, etc.


The dataset is publicly available at https://datacatalog.worldbank.org/dataset/climate-change-data and licenced under the <a href="https://datacatalog.worldbank.org/public-licenses#cc-by">Creative Commons Attribution 4.0 International license</a>.

***


## 1. Notebook Setup
Import all needed libraries:

In [1]:
import pandas as pd
import numpy as np

The complete Climate Change Dataset is imported into a pandas DataFrame from the downloaded file "climate_change_download_0.xls":

In [2]:
# define the file name and the data sheet
orig_data_file = r"climate_change_download_0.xls"
data_sheet = "Data"

# read the data from the excel file to a pandas DataFrame
data_orig = pd.read_excel(io=orig_data_file, sheet_name=data_sheet)

***

## 2. Global data overview

A global overview of the imported data yields the following insights:

In [3]:
print("Shape of the original dataset:")
data_orig.shape

Shape of the original dataset:


(13512, 28)

In [4]:
print("Available columns:")
data_orig.columns

Available columns:


Index(['Country code', 'Country name',  'Series code',  'Series name',
              'SCALE',     'Decimals',           1990,           1991,
                 1992,           1993,           1994,           1995,
                 1996,           1997,           1998,           1999,
                 2000,           2001,           2002,           2003,
                 2004,           2005,           2006,           2007,
                 2008,           2009,           2010,           2011],
      dtype='object')

In [5]:
print("Column data types:")
data_orig.dtypes

Column data types:


Country code    object
Country name    object
Series code     object
Series name     object
SCALE           object
Decimals        object
1990            object
1991            object
1992            object
1993            object
1994            object
1995            object
1996            object
1997            object
1998            object
1999            object
2000            object
2001            object
2002            object
2003            object
2004            object
2005            object
2006            object
2007            object
2008            object
2009            object
2010            object
2011            object
dtype: object

In [6]:
print("Overview of the first 5 rows:")
data_orig.head()

Overview of the first 5 rows:


Country code Country name     Series code  \
0          ABW        Aruba  AG.LND.EL5M.ZS   
1          ADO      Andorra  AG.LND.EL5M.ZS   
2          AFG  Afghanistan  AG.LND.EL5M.ZS   
3          AGO       Angola  AG.LND.EL5M.ZS   
4          ALB      Albania  AG.LND.EL5M.ZS   

                           Series name SCALE Decimals      1990 1991 1992  \
0  Land area below 5m (% of land area)     0        1   29.5748   ..   ..   
1  Land area below 5m (% of land area)     0        1         0   ..   ..   
2  Land area below 5m (% of land area)     0        1         0   ..   ..   
3  Land area below 5m (% of land area)     0        1  0.208235   ..   ..   
4  Land area below 5m (% of land area)     0        1   4.96788   ..   ..   

  1993  ... 2002 2003 2004 2005 2006 2007 2008 2009 2010 2011  
0   ..  ...   ..   ..   ..   ..   ..   ..   ..   ..   ..   ..  
1   ..  ...   ..   ..   ..   ..   ..   ..   ..   ..   ..   ..  
2   ..  ...   ..   ..   ..   ..   ..   ..   ..   ..   ..   ..  
3   ..  ...   ..   ..   ..   ..   ..   ..   ..   ..   ..   ..  
4   ..  ...   ..   ..   ..   ..   ..   ..   ..   ..   ..   ..  

[5 rows x 28 columns]

In [7]:
print("Descriptive statistics of the columns:")
data_orig.describe()

Descriptive statistics of the columns:


Country code Country name        Series code  \
count         13512        13512              13512   
unique          233          233                 58   
top             BOL   Uzbekistan  SE.ENR.PRSC.FM.ZS   
freq             58           58                233   

                                              Series name  SCALE  Decimals  \
count                                               13512  13512     13512   
unique                                                 58      2         3   
top     Public sector mgmt & institutions avg. (1-6 sc...      0         1   
freq                                                  233  10017      5823   

         1990   1991   1992   1993  ...   2002   2003   2004   2005   2006  \
count   10017  10017  10017  10017  ...  10017  10017  10017  10017  10017   
unique   4355   3398   3523   3583  ...   3877   3869   4007   4484   4008   
top        ..     ..     ..     ..  ...     ..     ..     ..     ..     ..   
freq     5163   6520   6364   6300  ...   5960   5974   5792   4933   5781   

         2007   2008   2009   2010   2011  
count   10017  10017  10017  10017  12382  
unique   4047   4080   3506   2164   1434  
top        ..     ..     ..     ..     ..  
freq     5769   5414   6256   7685  10244  

[4 rows x 28 columns]

In order to better understand the nature of the columns "Series code", "Series name', "SCALE" and "Decimals", it is necessary to examine their values.

The following snippet prints the contents of the column *'Series name'*:

In [8]:
data_orig['Series name'].unique()

array(['Land area below 5m (% of land area)',
       'Agricultural land under irrigation (% of total ag. land)',
       'Cereal yield (kg per hectare)',
       'Foreign direct investment, net inflows (% of GDP)',
       'Access to electricity (% of total population)',
       'Energy use per units of GDP (kg oil eq./$1,000 of 2005 PPP $)',
       'Energy use per capita (kilograms of oil equivalent)',
       'CO2 emissions, total (KtCO2)',
       'CO2 emissions per capita (metric tons)',
       'CO2 emissions per units of GDP (kg/$1,000 of 2005 PPP $)',
       'Other GHG emissions, total (KtCO2e)',
       'Methane (CH4) emissions, total (KtCO2e)',
       'Nitrous oxide (N2O) emissions, total (KtCO2e)',
       'Annex-I emissions reduction target',
       'Disaster risk reduction progress score (1-5 scale; 5=best)',
       'GHG net emissions/removals by LUCF (MtCO2e)',
       'Hosted Clean Development Mechanism (CDM) projects',
       'Hosted Joint Implementation (JI) projects',
       'Av

Contents of the column *'Series code'*:

In [9]:
data_orig['Series code'].unique()

array(['AG.LND.EL5M.ZS', 'AG.LND.IRIG.AG.ZS', 'AG.YLD.CREL.KG',
       'BX.KLT.DINV.WD.GD.ZS', 'EG.ELC.ACCS.ZS', 'EG.USE.COMM.GD.PP.KD',
       'EG.USE.PCAP.KG.OE', 'EN.ATM.CO2E.KT', 'EN.ATM.CO2E.PC',
       'EN.ATM.CO2E.PP.GD.KD', 'EN.ATM.GHGO.KT.CE', 'EN.ATM.METH.KT.CE',
       'EN.ATM.NOXE.KT.CE', 'EN.CLC.AERT', 'EN.CLC.DRSK.XQ',
       'EN.CLC.GHGR.MT.CE', 'EN.CLC.HCDM', 'EN.CLC.HJIP',
       'EN.CLC.HPPT.MM', 'EN.CLC.ICER', 'EN.CLC.IERU', 'EN.CLC.MDAT.ZS',
       'EN.CLC.MMDT.C', 'EN.CLC.NAMA', 'EN.CLC.NAPA', 'EN.CLC.NCOM',
       'EN.CLC.PCAT.C', 'EN.CLC.PCCC', 'EN.CLC.PCHW', 'EN.CLC.PCPT.MM',
       'EN.CLC.RNET', 'EN.POP.EL5M.ZS', 'EN.URB.MCTY.TL.ZS',
       'ER.H2O.FWTL.ZS', 'ER.LND.PTLD.ZS', 'IC.BUS.EASE.XQ',
       'IE.PPI.ENGY.CD', 'IE.PPI.TELE.CD', 'IE.PPI.TRAN.CD',
       'IE.PPI.WATR.CD', 'IQ.CPA.PUBS.XQ', 'IS.ROD.PAVE.ZS',
       'NY.GDP.MKTP.CD', 'NY.GNP.PCAP.CD', 'SE.ENR.PRSC.FM.ZS',
       'SE.PRM.CMPT.ZS', 'SH.DYN.MORT', 'SH.H2O.SAFE.ZS',
       'SH.MED.NUMW.P3', 'S

Contents of the column *'SCALE'*:

In [10]:
data_orig['SCALE'].unique()

array([0, 'Text'], dtype=object)

Contents of the column *'Decimals'*:

In [11]:
data_orig['Decimals'].unique()

array([1, 0, 'Text'], dtype=object)

At this point, it is unclear what are the values marked with the label *'Text'* in the *'SCALE'* and *'Decimals'* columns. These are shown in the following tow outputs:

In [12]:
data_orig[data_orig['SCALE']=='Text']

Country code                    Country name  Series code  \
3029          ABW                           Aruba  EN.CLC.AERT   
3030          ADO                         Andorra  EN.CLC.AERT   
3031          AFG                     Afghanistan  EN.CLC.AERT   
3032          AGO                          Angola  EN.CLC.AERT   
3033          ALB                         Albania  EN.CLC.AERT   
3034          ARE            United Arab Emirates  EN.CLC.AERT   
3035          ARG                       Argentina  EN.CLC.AERT   
3036          ARM                         Armenia  EN.CLC.AERT   
3037          ASM                  American Samoa  EN.CLC.AERT   
3038          ATG             Antigua and Barbuda  EN.CLC.AERT   
3039          AUS                       Australia  EN.CLC.AERT   
3040          AUT                         Austria  EN.CLC.AERT   
3041          AZE                      Azerbaijan  EN.CLC.AERT   
3042          BDI                         Burundi  EN.CLC.AERT   
3043          BEL                         Belgium  EN.CLC.AERT   
3044          BEN                           Benin  EN.CLC.AERT   
3045          BFA                    Burkina Faso  EN.CLC.AERT   
3046          BGD                      Bangladesh  EN.CLC.AERT   
3047          BGR                        Bulgaria  EN.CLC.AERT   
3048          BHR                         Bahrain  EN.CLC.AERT   
3049          BHS                    Bahamas, The  EN.CLC.AERT   
3050          BIH          Bosnia and Herzegovina  EN.CLC.AERT   
3051          BLR                         Belarus  EN.CLC.AERT   
3052          BLZ                          Belize  EN.CLC.AERT   
3053          BMU                         Bermuda  EN.CLC.AERT   
3054          BOL                         Bolivia  EN.CLC.AERT   
3055          BRA                          Brazil  EN.CLC.AERT   
3056          BRB                        Barbados  EN.CLC.AERT   
3057          BRN               Brunei Darussalam  EN.CLC.AERT   
3058          BTN                          Bhutan  EN.CLC.AERT   
...           ...                             ...          ...   
7193          TGO                            Togo  EN.CLC.RNET   
7194          THA                        Thailand  EN.CLC.RNET   
7195          TJK                      Tajikistan  EN.CLC.RNET   
7196          TKM                    Turkmenistan  EN.CLC.RNET   
7197          TMP                     Timor-Leste  EN.CLC.RNET   
7198          TON                           Tonga  EN.CLC.RNET   
7199          TTO             Trinidad and Tobago  EN.CLC.RNET   
7200          TUN                         Tunisia  EN.CLC.RNET   
7201          TUR                          Turkey  EN.CLC.RNET   
7202          TUV                          Tuvalu  EN.CLC.RNET   
7203          TZA                        Tanzania  EN.CLC.RNET   
7204          UGA                          Uganda  EN.CLC.RNET   
7205          UKR                         Ukraine  EN.CLC.RNET   
7206          UMC             Upper middle income  EN.CLC.RNET   
7207          URY                         Uruguay  EN.CLC.RNET   
7208          USA                   United States  EN.CLC.RNET   
7209          UZB                      Uzbekistan  EN.CLC.RNET   
7210          VCT  St. Vincent and the Grenadines  EN.CLC.RNET   
7211          VEN                   Venezuela, RB  EN.CLC.RNET   
7212          VIR           Virgin Islands (U.S.)  EN.CLC.RNET   
7213          VNM                         Vietnam  EN.CLC.RNET   
7214          VUT                         Vanuatu  EN.CLC.RNET   
7215          WBG              West Bank and Gaza  EN.CLC.RNET   
7216          WLD                           World  EN.CLC.RNET   
7217          WSM                           Samoa  EN.CLC.RNET   
7218          YEM                     Yemen, Rep.  EN.CLC.RNET   
7219          ZAF                    South Africa  EN.CLC.RNET   
7220          ZAR                Congo, Dem. Rep.  EN.CLC.RNET   
7221          ZMB                          Za

In [13]:
data_orig[data_orig['Decimals']=='Text']

Country code                    Country name  Series code  \
3029          ABW                           Aruba  EN.CLC.AERT   
3030          ADO                         Andorra  EN.CLC.AERT   
3031          AFG                     Afghanistan  EN.CLC.AERT   
3032          AGO                          Angola  EN.CLC.AERT   
3033          ALB                         Albania  EN.CLC.AERT   
3034          ARE            United Arab Emirates  EN.CLC.AERT   
3035          ARG                       Argentina  EN.CLC.AERT   
3036          ARM                         Armenia  EN.CLC.AERT   
3037          ASM                  American Samoa  EN.CLC.AERT   
3038          ATG             Antigua and Barbuda  EN.CLC.AERT   
3039          AUS                       Australia  EN.CLC.AERT   
3040          AUT                         Austria  EN.CLC.AERT   
3041          AZE                      Azerbaijan  EN.CLC.AERT   
3042          BDI                         Burundi  EN.CLC.AERT   
3043          BEL                         Belgium  EN.CLC.AERT   
3044          BEN                           Benin  EN.CLC.AERT   
3045          BFA                    Burkina Faso  EN.CLC.AERT   
3046          BGD                      Bangladesh  EN.CLC.AERT   
3047          BGR                        Bulgaria  EN.CLC.AERT   
3048          BHR                         Bahrain  EN.CLC.AERT   
3049          BHS                    Bahamas, The  EN.CLC.AERT   
3050          BIH          Bosnia and Herzegovina  EN.CLC.AERT   
3051          BLR                         Belarus  EN.CLC.AERT   
3052          BLZ                          Belize  EN.CLC.AERT   
3053          BMU                         Bermuda  EN.CLC.AERT   
3054          BOL                         Bolivia  EN.CLC.AERT   
3055          BRA                          Brazil  EN.CLC.AERT   
3056          BRB                        Barbados  EN.CLC.AERT   
3057          BRN               Brunei Darussalam  EN.CLC.AERT   
3058          BTN                          Bhutan  EN.CLC.AERT   
...           ...                             ...          ...   
7193          TGO                            Togo  EN.CLC.RNET   
7194          THA                        Thailand  EN.CLC.RNET   
7195          TJK                      Tajikistan  EN.CLC.RNET   
7196          TKM                    Turkmenistan  EN.CLC.RNET   
7197          TMP                     Timor-Leste  EN.CLC.RNET   
7198          TON                           Tonga  EN.CLC.RNET   
7199          TTO             Trinidad and Tobago  EN.CLC.RNET   
7200          TUN                         Tunisia  EN.CLC.RNET   
7201          TUR                          Turkey  EN.CLC.RNET   
7202          TUV                          Tuvalu  EN.CLC.RNET   
7203          TZA                        Tanzania  EN.CLC.RNET   
7204          UGA                          Uganda  EN.CLC.RNET   
7205          UKR                         Ukraine  EN.CLC.RNET   
7206          UMC             Upper middle income  EN.CLC.RNET   
7207          URY                         Uruguay  EN.CLC.RNET   
7208          USA                   United States  EN.CLC.RNET   
7209          UZB                      Uzbekistan  EN.CLC.RNET   
7210          VCT  St. Vincent and the Grenadines  EN.CLC.RNET   
7211          VEN                   Venezuela, RB  EN.CLC.RNET   
7212          VIR           Virgin Islands (U.S.)  EN.CLC.RNET   
7213          VNM                         Vietnam  EN.CLC.RNET   
7214          VUT                         Vanuatu  EN.CLC.RNET   
7215          WBG              West Bank and Gaza  EN.CLC.RNET   
7216          WLD                           World  EN.CLC.RNET   
7217          WSM                           Samoa  EN.CLC.RNET   
7218          YEM                     Yemen, Rep.  EN.CLC.RNET   
7219          ZAF                    South Africa  EN.CLC.RNET   
7220          ZAR                Congo, Dem. Rep.  EN.CLC.RNET   
7221          ZMB                          Za

### Findings from the global overview

This global overview gives away the following facts about the available data:
* shape: 28 columns, 13512 rows
* all columns are of type "object" - neither numeric, nor string/text values
* A certain amount of missing values, denoted both as NaN (not a number values) and as the string ".."
* The rows marked as *'Text'* in the columns *'SCALE'* and *'Decimals'* do not contain any information, almost completely composed of NaN values
* The columns represent key values such as country, but also the corresponding years and the series code/name
* The columns *'Country name'*, *'Series code'*, *'SCALE'* and *'Decimals'* do not give any information and are therefore obsolete
* The column *'Series name'* contains the country-specific features required for the analysis
* The names of the features in the column *'Series name'* are clear but too long


***

## 3. Define the initial project goals

The first overview of the raw data allows to define initial goals and objectives of the machine learning project. These will be refined in the future as more information insight is gained from the data. However, this initial goal definition will help develop a strategy and organize the data cleaning, transformation and visualization.

The data series available can be summarized into the following country-specific parameter/feature categories:

* various emissions of greenhouse gases such as CO2, CH4, N2O, others
* population-specific parameters: population count, urban population, population growth, etc.
* country economic indicators: GDP, GNI, Foreign Direct Investment, etc.
* land-related parameters: cereal yield, agricultural land, Nationally terrestrial protected areas, etc.
* climate data: precipitations, national disasters, etc.
* energy use
* counts of certain types of medical personnel
* etc.

Such a dataset would suggest to investigate the influence of country-specific parameters such as economic parameres, population, energy use, land use and others on climate-related data or the factors affecting the climate like emissions, precipitations, etc.

**Initial goal of the machine learning project:** Analyze the relationships among these variable categories and evaluate the contribution of factors like country economy, energy use, land use, etc. on greenhouse gas emissions, precipitations, etc. Finally, develop a machine learning model capable of predicting climate-related data or emissions from the other country-specific parameters.

As more data insight will be gained with along the course of the project, the definition of these goals will be refined in more detail.

***

## 4. Data cleaning

### Organization of the data cleaning and transformation

The main aim of the data cleaning and transformation is to represent the features (the country parameters contained in the column *'Series name'*) as separate columns and to make each row identifiable by a country and a year. At the same time, it would make sense to transform the years into a single column.

Additionally, it is necessary to get rid of empty rows or columns and deal with the remaining cells with missing values.

For these purposes, the following tasks have to be undertaken:

1. Remove rows marked as "Text" in the "SCALE" and "Decimals" columns
2. Remove the unnecessary columns "Country name", "Series code", "SCALE", "Decimals"
3. Transform the ".." strings and emplty cells ("") into NaN values for easier recognission as missing values
4. Transform all data columns into a numerical data type
5. Rename the features in column "Series name"

### 4.1 Removing rows marked as "Text" in the "SCALE" and "Decimals" columns

In [14]:
# assign the data to a new DataFrame, which will be modified
data_clean = data_orig

print("Original number of rows:")
print(data_clean.shape[0])

# remove rows characterized as "Text" in the SCALE column
data_clean = data_clean[data_clean['SCALE']!='Text']

print("Current number of rows:")
print(data_clean.shape[0])

Original number of rows:
13512
Current number of rows:
10017


### 4.2 Removing the unnecessary columns "Country name", "Series code", "SCALE", "Decimals"

In [15]:
print("Original number of columns:")
print(data_clean.shape[1])

data_clean = data_clean.drop(['Country name', 'Series code', 'SCALE', 'Decimals'], axis='columns')

print("Current number of columns:")
print(data_clean.shape[1])

Original number of columns:
28
Current number of columns:
24


### 4.3 Transform the ".." strings and emplty cells ("") into NaN values for easier recognission as missing values

In [16]:
data_clean.iloc[:,2:] = data_clean.iloc[:,2:].replace({'':np.nan, '..':np.nan})

### 4.4 Transform all data columns into a numerical data type

In [17]:
data_clean2 = data_clean.applymap(lambda x: pd.to_numeric(x, errors='ignore'))
# Errors are ignored in order to avoid error messages about the first two columns, which don't need to be transformed
# into numeric type anyway

print("Print the column data types after transformation:")
data_clean2.dtypes

Print the column data types after transformation:


Country code     object
Series name      object
1990            float64
1991            float64
1992            float64
1993            float64
1994            float64
1995            float64
1996            float64
1997            float64
1998            float64
1999            float64
2000            float64
2001            float64
2002            float64
2003            float64
2004            float64
2005            float64
2006            float64
2007            float64
2008            float64
2009            float64
2010            float64
2011            float64
dtype: object

### 4.5 Rename the features in column "Series name"

The variable/feature names in the column *'Series name'* are clear, but too long and not practical to use in the code. In order to improve that, the most relevant feature names will be renamed with shorter labels as indicated in the following table:

<table>
    <tr>
        <td> <b>Variable name</b> </td>
        <td> <b>Description</b> </td>
        <td> <b>Unit</b> </td>
    </tr>
    <tr>
        <td> cereal_yield </td>
        <td> Cereal yield </td>
        <td> kg per hectare </td>
    </tr>
    <tr>
        <td> fdi_perc_gdp </td>
        <td> Foreign direct investment, net inflows </td>
        <td> % of GDP </td>
    </tr>
    <tr>
        <td> elec_access_perc </td>
        <td> Access to electricity </td>
        <td> % of total population </td>
    </tr>
    <tr>
        <td> en_per_gdp </td>
        <td> Energy use per units of GDP </td>
        <td> kg oil eq./\$1,000 of 2005 PPP \$ </td>
    </tr>
    <tr>
        <td> en_per_cap </td>
        <td> Energy use per capita </td>
        <td> kilograms of oil equivalent </td>
    </tr>
    <tr>
        <td> co2_ttl </td>
        <td> CO2 emissions, total </td>
        <td> KtCO2 </td>
    </tr>
    <tr>
        <td> co2_per_cap </td>
        <td> CO2 emissions, total </td>
        <td> metric tons </td>
    </tr>
    <tr>
        <td> co2_per_gdp </td>
        <td> CO2 emissions per units of GDP </td>
        <td> kg/\$1,000 of 2005 PPP \$ </td>
    </tr>
    <tr>
        <td> other_ghg_ttl </td>
        <td> Other GHG emissions, total </td>
        <td> KtCO2e </td>
    </tr>
    <tr>
        <td> ch4_ttl </td>
        <td> Methane (CH4) emissions, total </td>
        <td> KtCO2 </td>
    </tr>
    <tr>
        <td> n2o_ttl </td>
        <td> Nitrous oxide (N2O) emissions, total </td>
        <td> KtCO2 </td>
    </tr>
    <tr>
        <td> nat_emerg </td>
        <td> Droughts, floods, extreme temps </td>
        <td> % pop. avg. 1990-2009 </td>
    </tr>
    <tr>
        <td> pop_urb_aggl_perc </td>
        <td> Population in urban agglomerations >1million </td>
        <td> % </td>
    </tr>
    <tr>
        <td> prot_area_perc </td>
        <td> Nationally terrestrial protected areas </td>
        <td> % of total land area </td>
    </tr>
    <tr>
        <td> gdp </td>
        <td> Gross Domestic Product (GDP) </td>
        <td> \$ </td>
    </tr>
    <tr>
        <td> gni_per_cap </td>
        <td> GNI per capita </td>
        <td> Atlas \$ </td>
    </tr>
    <tr>
        <td> under_5_mort_rate </td>
        <td> Under-five mortality rate </td>
        <td> per 1,000 </td>
    </tr>
    <tr>
        <td> pop_growth_perc </td>
        <td> Population growth </td>
        <td> annual % </td>
    </tr>
    <tr>
        <td> pop </td>
        <td> Population </td>
        <td> 1 </td>
    </tr>
    <tr>
        <td> urb_pop_growth_perc </td>
        <td> Urban population growth </td>
        <td> annual % </td>
    </tr>
    <tr>
        <td> urb_pop </td>
        <td> Urban population </td>
        <td> 1 </td>
    </tr>
    
</table>

In [18]:
# define shorter names corresponding to most relevant variables in a dictionary
chosen_vars = {'Cereal yield (kg per hectare)': 'cereal_yield',
               'Foreign direct investment, net inflows (% of GDP)': 'fdi_perc_gdp',
               'Access to electricity (% of total population)': 'elec_access_perc',
               'Energy use per units of GDP (kg oil eq./$1,000 of 2005 PPP $)': 'en_per_gdp',
               'Energy use per capita (kilograms of oil equivalent)': 'en_per_cap',
               'CO2 emissions, total (KtCO2)': 'co2_ttl',
               'CO2 emissions per capita (metric tons)': 'co2_per_cap',
               'CO2 emissions per units of GDP (kg/$1,000 of 2005 PPP $)': 'co2_per_gdp',
               'Other GHG emissions, total (KtCO2e)': 'other_ghg_ttl',
               'Methane (CH4) emissions, total (KtCO2e)': 'ch4_ttl',
               'Nitrous oxide (N2O) emissions, total (KtCO2e)': 'n2o_ttl',
               'Droughts, floods, extreme temps (% pop. avg. 1990-2009)': 'nat_emerg',
               'Population in urban agglomerations >1million (%)': 'pop_urb_aggl_perc',
               'Nationally terrestrial protected areas (% of total land area)': 'prot_area_perc',
               'GDP ($)': 'gdp',
               'GNI per capita (Atlas $)': 'gni_per_cap',
               'Under-five mortality rate (per 1,000)': 'under_5_mort_rate',
               'Population growth (annual %)': 'pop_growth_perc',
               'Population': 'pop',
               'Urban population growth (annual %)': 'urb_pop_growth_perc',
               'Urban population': 'urb_pop'
                }

# rename all variables in the column "Series name" with comprehensible shorter versions
data_clean2['Series name'] = data_clean2['Series name'].replace(to_replace=chosen_vars)

***

## 5. Data frame transformation

This is how the current data frame looks like:

In [19]:
data_clean2.head()

Country code                          Series name       1990  1991  1992  \
0          ABW  Land area below 5m (% of land area)  29.574810   NaN   NaN   
1          ADO  Land area below 5m (% of land area)   0.000000   NaN   NaN   
2          AFG  Land area below 5m (% of land area)   0.000000   NaN   NaN   
3          AGO  Land area below 5m (% of land area)   0.208235   NaN   NaN   
4          ALB  Land area below 5m (% of land area)   4.967875   NaN   NaN   

   1993  1994  1995  1996  1997  ...  2002  2003  2004  2005  2006  2007  \
0   NaN   NaN   NaN   NaN   NaN  ...   NaN   NaN   NaN   NaN   NaN   NaN   
1   NaN   NaN   NaN   NaN   NaN  ...   NaN   NaN   NaN   NaN   NaN   NaN   
2   NaN   NaN   NaN   NaN   NaN  ...   NaN   NaN   NaN   NaN   NaN   NaN   
3   NaN   NaN   NaN   NaN   NaN  ...   NaN   NaN   NaN   NaN   NaN   NaN   
4   NaN   NaN   NaN   NaN   NaN  ...   NaN   NaN   NaN   NaN   NaN   NaN   

   2008  2009  2010  2011  
0   NaN   NaN   NaN   NaN  
1   NaN   NaN   NaN   NaN  
2   NaN   NaN   NaN   NaN  
3   NaN   NaN   NaN   NaN  
4   NaN   NaN   NaN   NaN  

[5 rows x 24 columns]

As mentioned previously, it is necessary to transform the data frame into a suitable format - the features from the *'Series name'* column into separate columns and the years into a single column. This is done by picking the corresponding values of each feature together with their countries and years, melting these into a single column and then combine them into a new data frame with all features (basing on the same countries and years):

In [20]:
# save the short feature names into a list of strings
chosen_cols = list(chosen_vars.values())

# define an empty list, where sub-dataframes for each feature will be saved
frame_list = []

# iterate over all chosen features
for variable in chosen_cols:
    
    # pick only rows corresponding to the current feature
    frame = data_clean2[data_clean2['Series name'] == variable]
    
    # melt all the values for all years into one column and rename the columns correspondingly
    frame = frame.melt(id_vars=['Country code', 'Series name']).rename(columns={'Country code': 'country', 'variable': 'year', 'value': variable}).drop(['Series name'], axis='columns')
    
    # add the melted dataframe for the current feature into the list
    frame_list.append(frame)


# merge all sub-frames into a single dataframe, making an outer binding on the key columns 'country','year'
from functools import reduce
all_vars = reduce(lambda left, right: pd.merge(left, right, on=['country','year'], how='outer'), frame_list)

After this transformation, the new data frame has the following layout:

In [21]:
all_vars.head()

country  year  cereal_yield  fdi_perc_gdp  elec_access_perc  en_per_gdp  \
0     ABW  1990           NaN           NaN               NaN         NaN   
1     ADO  1990           NaN           NaN               NaN         NaN   
2     AFG  1990        1200.6           NaN               NaN         NaN   
3     AGO  1990         320.9     -3.263096               NaN  172.978279   
4     ALB  1990        2794.3           NaN               NaN  206.751128   

   en_per_cap   co2_ttl  co2_per_cap  co2_per_gdp  ...  nat_emerg  \
0         NaN  1840.834    29.620641          NaN  ...        NaN   
1         NaN       NaN          NaN          NaN  ...        NaN   
2         NaN  2676.910     0.140715          NaN  ...        NaN   
3  569.242999  4429.736     0.428613   130.244398  ...        NaN   
4  809.215612  7488.014     2.276350   581.597641  ...        NaN   

   pop_urb_aggl_perc  prot_area_perc           gdp  gni_per_cap  \
0                NaN        0.105547           NaN          NaN   
1                NaN        5.559145  1.028989e+09      17440.0   
2           6.739396        0.432993           NaN          NaN   
3          15.171061       12.399821  1.026019e+10        740.0   
4                NaN        4.336615  2.101625e+09        680.0   

   under_5_mort_rate  pop_growth_perc         pop  urb_pop_growth_perc  \
0                NaN         1.820254     62147.0             1.780501   
1                8.8         2.945970     52773.0             2.840429   
2              208.7         2.082624  19023678.0             3.492189   
3              243.0         2.771320  10335052.0             6.505966   
4               41.1         1.027427   3289483.0             1.744276   

       urb_pop  
0    31259.941  
1    49976.031  
2  3481333.074  
3  3834304.292  
4  1197371.812  

[5 rows x 23 columns]

***

## 6. Remove the remaining missing values in an optimal way

Although some columns and rows with empty cells have already been deleted, there are still remaining missing values:

In [22]:
print("check the amount of missing values in each column")
all_vars.isnull().sum()

check the amount of missing values in each column


country                   0
year                      0
cereal_yield           1377
fdi_perc_gdp           1111
elec_access_perc       5027
en_per_gdp             2082
en_per_cap             1956
co2_ttl                1143
co2_per_cap            1146
co2_per_gdp            1557
other_ghg_ttl          4542
ch4_ttl                4526
n2o_ttl                4526
nat_emerg              4958
pop_urb_aggl_perc      2582
prot_area_perc          726
gdp                     779
gni_per_cap            1013
under_5_mort_rate       716
pop_growth_perc         278
pop                     252
urb_pop_growth_perc     490
urb_pop                 467
dtype: int64

The objective here is to remove as many empty cells as possible, while preserving the highest possible amount of columns and rows. Missing values can originate from missing data for certain years, countries or features (columns). For this reason, it is necessary to filter the data by these three categories and remove rows or columns with NaNs by starting with the row/column with the highest amount of missing cells. This way to clean the dataset will preserve the maximal amount of possibly valuable information in columns and rows rather than just deleting all rows with any number of missing values.

Since there are multiple appearances of the unique values of countries and years, it is necessary to count the number of NaN values for each unique year and country.

### 6.1 Filtering the years by missing values

Checking the amount of missing values for each year:

In [23]:
all_vars_clean = all_vars

#define an array with the unique year values
years_count_missing = dict.fromkeys(all_vars_clean['year'].unique(), 0)
for ind, row in all_vars_clean.iterrows():
    years_count_missing[row['year']] += row.isnull().sum()

# sort the years by missing values
years_missing_sorted = dict(sorted(years_count_missing.items(), key=lambda item: item[1]))

# print the missing values for each year
print("missing values by year:")
for key, val in years_missing_sorted.items():
    print(key, ":", val)

missing values by year:
2005 : 1189
2000 : 1273
1995 : 1317
1990 : 1427
2007 : 1631
2006 : 1633
2004 : 1646
2008 : 1708
2003 : 1714
2002 : 1715
2001 : 1718
1999 : 1729
1998 : 1739
1997 : 1746
1996 : 1756
1994 : 1781
1993 : 1792
1992 : 1810
1991 : 1921
2009 : 2078
2010 : 3038
2011 : 4893


The purpose of the filtering is to delete rows with a significant amount of missing values for certain countries without removing too many years. So it is important to choose the proper limit for NaN values allowed per year. The previous output suggests to pick the years between 1991 and 2008 for the further analysis:

In [24]:
print("number of missing values in the whole dataset before filtering the years:")
print(all_vars_clean.isnull().sum().sum())
print("number of rows before filtering the years:")
print(all_vars_clean.shape[0])

# filter only rows for years between 1991 and 2008 (having less missing values)
all_vars_clean = all_vars_clean[(all_vars_clean['year'] >= 1991) & (all_vars_clean['year'] <= 2008)]

print("number of missing values in the whole dataset after filtering the years:")
print(all_vars_clean.isnull().sum().sum())
print("number of rows after filtering the years:")
print(all_vars_clean.shape[0])

number of missing values in the whole dataset before filtering the years:
41254
number of rows before filtering the years:
5126
number of missing values in the whole dataset after filtering the years:
29818
number of rows after filtering the years:
4194


### 6.2 Filtering the countries by missing values

The same procedure is applied to the filtering of countries with missing values. The following snippet shows the number of NaNs for each country.

In [25]:
# check the amount of missing values by country

# define an array with the unique country values
countries_count_missing = dict.fromkeys(all_vars_clean['country'].unique(), 0)

# iterate through all rows and count the amount of NaN values for each country
for ind, row in all_vars_clean.iterrows():
    countries_count_missing[row['country']] += row.isnull().sum()

# sort the countries by missing values
countries_missing_sorted = dict(sorted(countries_count_missing.items(), key=lambda item: item[1]))

# print the missing values for each country
print("missing values by country:")
for key, val in countries_missing_sorted.items():
    print(key, ":", val)

missing values by country:
AGO : 81
ARG : 81
AUS : 81
AUT : 81
BGD : 81
BGR : 81
BOL : 81
BRA : 81
CAN : 81
CHE : 81
CHL : 81
CHN : 81
CIV : 81
CMR : 81
COG : 81
COL : 81
CRI : 81
DEU : 81
DNK : 81
DOM : 81
ECU : 81
EGY : 81
EMU : 81
ESP : 81
FIN : 81
FRA : 81
GBR : 81
GHA : 81
GTM : 81
HND : 81
HUN : 81
IDN : 81
IND : 81
IRL : 81
ISR : 81
ITA : 81
JOR : 81
JPN : 81
KEN : 81
KOR : 81
LAC : 81
LMC : 81
LMY : 81
MAR : 81
MEX : 81
MIC : 81
MNA : 81
MOZ : 81
MYS : 81
NGA : 81
NLD : 81
NZL : 81
PAK : 81
PAN : 81
PER : 81
PHL : 81
PRT : 81
PRY : 81
ROM : 81
SAS : 81
SAU : 81
SDN : 81
SEN : 81
SLV : 81
SWE : 81
SYR : 81
TGO : 81
THA : 81
TUR : 81
TZA : 81
UMC : 81
URY : 81
USA : 81
VEN : 81
VNM : 81
ZAF : 81
ZMB : 81
GRC : 82
POL : 82
YEM : 82
ZAR : 82
DZA : 84
ETH : 84
LIC : 84
SSA : 84
WLD : 84
ARE : 85
ECA : 85
RUS : 86
UKR : 86
ARM : 87
BLR : 87
UZB : 87
KAZ : 88
CZE : 89
IRN : 89
BEL : 90
AZE : 91
GEO : 92
LBN : 92
HTI : 94
NIC : 96
BEN : 99
BWA : 99
CYP : 99
GAB : 99
HIC : 99
JAM : 99
K

This output would suggest to remove rows for countries with more than 90 missing values:

In [26]:
print("number of missing values in the whole dataset before filtering the countries:")
print(all_vars_clean.isnull().sum().sum())
print("number of rows before filtering the countries:")
print(all_vars_clean.shape[0])


# filter only rows for countries with less than 90 missing values
countries_filter = []
for key, val in countries_missing_sorted.items():
    if val<90:
        countries_filter.append(key)

all_vars_clean = all_vars_clean[all_vars_clean['country'].isin(countries_filter)]

print("number of missing values in the whole dataset after filtering the countries:")
print(all_vars_clean.isnull().sum().sum())
print("number of rows after filtering the countries:")
print(all_vars_clean.shape[0])

number of missing values in the whole dataset before filtering the countries:
29818
number of rows before filtering the countries:
4194
number of missing values in the whole dataset after filtering the countries:
7854
number of rows after filtering the countries:
1728


### 6.3 Checking the features (columns) for missing values

The NaN values count in each column is:

In [27]:
all_vars_clean.isnull().sum()

country                   0
year                      0
cereal_yield             10
fdi_perc_gdp             17
elec_access_perc       1728
en_per_gdp                0
en_per_cap                0
co2_ttl                   9
co2_per_cap               9
co2_per_gdp               9
other_ghg_ttl          1446
ch4_ttl                1440
n2o_ttl                1440
nat_emerg              1728
pop_urb_aggl_perc         0
prot_area_perc            0
gdp                       2
gni_per_cap              16
under_5_mort_rate         0
pop_growth_perc           0
pop                       0
urb_pop_growth_perc       0
urb_pop                   0
dtype: int64

After having filtered the years and countries with the most missing values, the features *elec_access_perc*, *other_ghg_ttl*, *ch4_ttl*, *n20_ttl* and *nat_emerg* still contain a significant number of missing values, removing which will reduce the amount of observations very drastically. Therefore, the next step is to remove these columns:

In [28]:
# remove features with more than 20 missing values

from itertools import compress

# create a boolean mapping of features with more than 20 missing values
vars_bad = all_vars_clean.isnull().sum()>20

# remove the columns corresponding to the mapping of the features with many missing values
all_vars_clean2 = all_vars_clean.drop(compress(data = all_vars_clean.columns, selectors = vars_bad), axis='columns')

print("Remaining missing values per column:")
print(all_vars_clean2.isnull().sum())

Remaining missing values per column:
country                 0
year                    0
cereal_yield           10
fdi_perc_gdp           17
en_per_gdp              0
en_per_cap              0
co2_ttl                 9
co2_per_cap             9
co2_per_gdp             9
pop_urb_aggl_perc       0
prot_area_perc          0
gdp                     2
gni_per_cap            16
under_5_mort_rate       0
pop_growth_perc         0
pop                     0
urb_pop_growth_perc     0
urb_pop                 0
dtype: int64


Removing the rows with the remainin missing values will not impair the size of the dataset significantly, so these rows will be deleted:

In [29]:
# delete rows with any number of missing values
all_vars_clean3 = all_vars_clean2.dropna(axis='rows', how='any')

print("Remaining missing values per column:")
print(all_vars_clean3.isnull().sum())

print("Final shape of the cleaned dataset:")
print(all_vars_clean3.shape)

Remaining missing values per column:
country                0
year                   0
cereal_yield           0
fdi_perc_gdp           0
en_per_gdp             0
en_per_cap             0
co2_ttl                0
co2_per_cap            0
co2_per_gdp            0
pop_urb_aggl_perc      0
prot_area_perc         0
gdp                    0
gni_per_cap            0
under_5_mort_rate      0
pop_growth_perc        0
pop                    0
urb_pop_growth_perc    0
urb_pop                0
dtype: int64
Final shape of the cleaned dataset:
(1700, 18)


***

## 7. Export of the cleaned data frame to a file

Now that the dataset has been rearranged and cleaned of missing values, it can be exported to a csv file (without the row index) for further analysis:

In [30]:
# export the clean dataframe to a csv file
all_vars_clean3.to_csv('data_cleaned.csv', index=False)

The further stages of the project are Data Visualization and Predictive Analysis, which can be found in the corresponding notebooks.

***

Copyright (c) 2020 Vladislav Todorov